libgedathon/docs/gsym-snap-grid.1.in
.TH gsym-snap-grid 1 "@DATE@" "gEDA Project" @VERSION@

.SH NAME
gsym-snap-grid \- A gEDA-gaf Symbol File Utility

.SH SYNOPSIS
.B gsym-snap-grid
.RI  [ INFORMATION ] | [ OPTIONS ] [ DIRECTIVES ] [INPUT] [OUTPUT]

.SH DESCRIPTION
.B gsym-snap-grid
Round X-Y coordinates of attributes in a symbol file to multibles of 25 or a specified value.

.SH QUICK
.B usage: [-h] | [Options] -i <inputfile> [[-o] <outputfile> ]

ex. 1) gsym-snap-grid incomming

Example: Snap all attribute coordinates values in all symbol files in the current directory and in all subdirectories to the nearest multiple of 50 and         report modified values:

    gsym-snap-grid.py --verbose -R -m 50

.SH INFORMATION
Use of the following options displays information related to gsym-snap-grid
but does not inspect or alter any symbol files:
.IP "-h, --help"
Show information about the advanced usage of this command.
.IP "-V, --version"
Displays software version information.

.SH OPTIONS
The following options can be used to modify the default behavior of gsym-snap-grid

.IP "-d, --down"
"Force rounding down, the default is to round to nearest multible of moduli.
.IP "-u, --up"
Force rounding up, the default is to round to nearest multible of moduli.
.IP "-m #"
Round to moduli specified by #, default is 25.
.IP "-x"
Only round X coordinates, mutually exclusive with -y
.IP "-y"
Only round Y coordinates, mutually exclusive with -x

.SH DIRECTIVES
.IP "-v, --verbose"
Verbose mode.  Used in both archive and extract mode. Spews lots of info about what the prog is doing.
.IP "-R, --Recursive"
Process all symbol files in the current and all subordinated directories.
.IP "-i, --input <input file>"
Specify the name of the input sym file. Wild cards are accepted, the default is *.sym.
.IP "-o, --output <output file>"
Specifies the name of the output sym file. When the output file name is not specified the input file will be over written

.SH HOMEPAGE
http://www.gpleda.org/

.SH BUGS
You can report bugs at the Launchpad site of gsym-snap-grid:
https://bugs.launchpad.net/geda/gsym-snap-grid/+filebug

.SH AUTHOR
Wiley Edward Hill < wileyhill dot gmail dot com >

.SH SEE ALSO
.BR gsym-attribute (1) gsym-set-font-size (1) gsym-set-visibility (1)

.SH COPYRIGHT
.nf
Copyright \(co 2015 Wiley Edward Hill.  License GPLv3+: GNU GPL
version 3.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2014 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 2.0.9 - Rampage Edition

* Gschem has improved performance and improved memory management. The main
  menu has been revised for better organization and extended with improved
  access to simulation netlist generators. New features have been added
  with emphasis on user configurability.

  A new Session system manages collections of schematics and or symbols.
  Sessions can be invoked at startup via a new command-line option and
  offers a configurable startup dialog. The active session name can
  optionally be displayed in the title bar.

  Users can now elect to show the full path in the title bar or just the basic
  filename as before. File names are prefixed with an asterisk when ever the
  current page has been modified and not saved.

  A new Component Properties dialog allows users to change components with
  out having to edit the schematic file with an editor. The Component
  Properties dialog is also used for editing symbols, allowing users to set
  or edit standard attributes for symbols from a single access point.

  Grids systems have been completely revised and feature user configurable
  dual color mesh and dots. Colors for the minor and major step can be set
  separately for each grid type. The alpha transparency for grid types are
  also user configurable. The settings can be adjusted to produce the same
  "dull gray" lines and dots as before but adding colors can improve the
  contrast and usability of the grids.

  The File Chooser dialog used for selecting and saving documents has been
  replaced. The new dialog retains position and geometry settings like other
  gschem dialogs. The File Chooser dialog also retains user filter preferences
  between sessions.

* Additonal Symbols, including symbols for newer more modern semiconductors.

* Libgeda API has been expanded to provide more functionality and has
  improved stability. The API is cleaner and the Doxygen documentation
  has been improved.

* Libgedacairo offers improved image quality and better Doxygen documentation.

* Libgedathon now accept either a string or an integer for the pin number
  argument. The Doxygen documentation has been improved.

* Libgedauio has more Widgets and expanded Doxygen documentation. The
  new File Chooser implemented in gschem is a new Libgedauio dialog.
  The library's build system is better organized to accommodate future
  expansion. Other new widgets include custom Combo boxes and TreeView.
  The Doxygen documentation has been expanded.

* The build system is compatible with Automake 1.9, the previous version
  required Automake 1.11 or newer but this is not available on some older
  systems.

  Numerous Bugs and annoyances resolved!

Notable changes in gEDA/gaf 2.0.8 - Rampage Edition

* Gschem has improved performances, additional toolbars, improved stability,
  enhanced appearance and improved memory management. The main menu has been
  revised to improve clarity and organization. New features have been added
  such a Invert Selection, Zoom-To Selection,

* Libgeda extensively revised to provide a more generic API. The library
  has improved performance from previous versions and uses less memory,
  The was mostly achieved by eliminating the needless passing of parameters.
  All Object types are now object oriented to improve memory management
  by off-loading responsibility to glib. The API has been extended to
  provide more functionality for applications. The configuration storage
  systems provided by Libgeda have been revised, mostly the new system,
  which no longer utilizes gnome based virtual file systems, aka gvfs. The
  use of GFile exponentially increased dependencies and was eliminated.

* Symbol libraries significantly expanded.

* A new library, Libgedathon provides an API for Python scripts.

* A new library, Libgedawidget provides an "Gtk" like widgets to enhance
  geda-gaf applications and extend support of code obsoleted by Gtk
  developers.

* The build system has been "heavily" modified and the source code
  completely reorganized. New macros provide better organization and
  greater flexibility.

* New install scripts, geda-install and geda-install-bin facilitate
  installation from local builds. (Note that geda-install-bin replaces
  ginstall_local)

Notable changes in gEDA/gaf 2.0.7.8 - Rampage Edition

* The Coordinates Dialog allows direct entering of World Coordinates.
  If an XY coordinate is "entered" when not in an Action, the cursor
  will be positioned over the point so that "hot-keys" can be used,
  to begin an action starting from an exact point. Parentheses for
  the coordinates is optional.

* Gschem now supports bidirectional Drag & Drop. Contents of Schematic
  drawings can be "dragged" from one gschem window to any other instance
  of gschem. The target can also be the World Entry of the Coordinates
  Dialog of the same or another instances of gschem. Dropping schematic
  objects on the World Entry requires that a value be entered immediately
  afterward for the dropped object to be inserted into the drawing at the
  specified point. Dropping a ".sch" file results in the file being loaded
  into new window. The results of dropping a ".sym" file depends on where
  the file is located. If the symbol file is in a Library source directory
  then the symbol is inserted into the current Schematic, otherwise the
  symbol file is opened in a new window for editing as a Symbol definition.

* "Save As" can now optionally add the ".sch" or ".sym" file extension.
  based on the filter selection.

* Generated color images file can now be inverted. User can choose between

libgedathon/scripts/gsym-snap-grid.py
#! /usr/bin/env python

# Set Visibility of Attributes (For now just to Defaults)

import os, sys, glob
from geda import geda
from geda.constants import *

#---------------------------------------------------------------------------
Version="0.1.0"
#---------------------------------------------------------------------------
VerboseMode=False
#---------------------------------------------------------------------------

#  This is the help string.
Usage =\
"""
gsym-snap-grid -- A gEDA-gaf Symbol File Utility
                  Round X-Y coordinates of attributes in a symbol file
                  to multibles of 25 or a specified value.

Usage: gsym-snap-grid [Options] -i <inputfile> [[-o] <outputfile> ]
"""

Help =\
"""
Options:

  -R, --Recursive -- Process all symbol files in the current and all subordinated directories.
  -v, --verbose   -- Verbose mode.  Used in both archive and extract mode.
                     Spews lots of info about what the prog is doing.

  The following options can be used to modify the default behavior of gsym-snap-grid

  -u, --up   Force rounding down, defaults to nearest.
  -d, --down Force rounding up, defaults to nearest.

  -m # Round to moduli specified by #, default is 25.
  -x   Only round X coordinates, mutually exclusive with -y
  -y   Only round Y coordinates, mutually exclusive with -x

Example: Snap all attribute coordinates values in all symbol files in the current
         directory and in all subdirectories to the nearest multiple of 50 and
         report modified values:

    gsym-snap-grid.py --verbose -R -m 50

Copyright (C) 2015 by Wiley Edward Hill. Released under GPL Version 3.

"""

#-------------------------------- Classes ----------------------------------
class ProgramParameters:
    """
    This class holds info about the environment and cmd line args passed to the
    program.  It has only one method:  the constructor, which gets the args
    and fills out the public vars.  The public vars are:
    """
    def __init__(self):

        global Version

        """
        Constructor: parse through cmd line args and fill out vars. The values
        here are the defaults
        """
        self.RecursiveMode   = False
        self.VerboseMode     = False
        self.InputFiles      = []
        self.UserDir         = os.path.abspath(os.getcwd())

        self.rnd_up   = False
        self.rnd_down = False
        self.fix_x    = True
        self.fix_y    = True
        self.moduli   = 25

        if len(sys.argv) > 1:

            mflag = False

            for arg in sys.argv[1:]:       # Skip OurSelf

                if mflag:
                    self.moduli = int(arg)
                    mflag = False
                    continue

                if arg in ('-V', '--version'):
                    print Version
                    sys.exit(0)

                if arg in ('-h', '--help'):
                    print Usage
                    print Help
                    sys.exit(0)

                if arg in ('-v', '--verbose'):
                    self.VerboseMode = True
                    continue

                if arg in ('-R', '--Recursive'):
                    self.RecursiveMode = True
                    continue

                if arg in ('-d', '--down'):
                    self.rnd_down = True
                    continue

                if arg in ('-u', '--up'):
                    self.rnd_up = True
                    continue

                if arg in ('-x', '-X'):
                    self.fix_y = False
                    continue

                if arg in ('-y', '-Y'):
                    self.fix_x = False
                    continue

                # The set to 10 points group
                if arg in ('-m', '-M', '--moduli'):
                    mflag = True
                    continue

                self.InputFiles.append(arg)

#---------------------------------------------------------------------------
def round_down(num, divisor):
    return num - (num%divisor)
#---------------------------------------------------------------------------
def round_up(num, divisor):
    return num + (num%divisor)
#---------------------------------------------------------------------------

def SnapText(Options, Text):
    modified = False
    if Options.fix_x:
        if not Text.x % Options.moduli == 0:
            modified = modified + 1
            low  = Text.x - Text.x % Options.moduli
            high = Text.x + Options.moduli - Text.x % Options.moduli
            if Options.VerboseMode:
                 print "Text: " + Text.string + " " + str(Text.x) + " is off X grid"
            if Options.rnd_down:
                Text.x = low
            elif Options.rnd_up:
                Text.x = high
            elif Text.x - low <= high - Text.x:
                Text.x = low
            else:
                Text.x = high

    if Options.fix_y:
        if not Text.y % Options.moduli == 0:
            modified = modified + 1
            low  = Text.y - Text.y % Options.moduli
            high = Text.y + Options.moduli - Text.y % Options.moduli
            if Options.VerboseMode:
                 print "Text: " + Text.string + " " + str(Text.y) + " is off Y grid"
            if Options.rnd_down:
                Text.y = low
            elif Options.rnd_up:
                Text.y = high
            elif Text.y - low <= high - Text.y:
                Text.y = low
            else:
                Text.y = high

    if modified and Options.VerboseMode:
        print "New postion: (" + str(Text.x) + "," + str(Text.y) + ")"

    return modified

#---------------------------------------------------------------------------

def ProcessSymbol(Options, File):

    symbol = geda.open_page(File)

    if Options.VerboseMode:
        print "Processing: " + symbol.filename()

    objects  = geda.get_objects(symbol)
    modified = 0
    """
      Loop thru all objects and get list of detached text attribute then loop
      again and skip if in list, Sub object (attrinutes) are to be handled by
      the objects. This is so that detached text gets added to the symbol sym
      groups.
    """
    for capsule in objects:
        object = geda.get_object(capsule)          # Retrieve object from GedaCapsule
        if capsule.type == OBJ_TEXT:
            if SnapText(Options, object):
                modified = modified + 1

    if modified:
        if Options.VerboseMode:
            print "Made " + str(modified) + " modifications to " + symbol.filename()
        symbol.save() # Write the symbol to storage

    # Close the file
    symbol.close()

#---------------------------------------------------------------------------

def RecursiveGlob(path, *exts):
	""" Glob recursively a directory and all subdirectories for multiple file extensions
    Note: Glob is case-insensitive, i. e. for '\*.jpg' you will get files ending
    with .jpg and .JPG

    Parameters
    ----------
    path : str
        A directory name
    exts : tuple
        File extensions to glob for

    Returns
    -------
    files : list
        list of files matching extensions in exts in path and subfolders
	"""
	dirs = [a[0] for a in os.walk(path)]
	f_filter = [d+e for d in dirs for e in exts]
	return [f for files in [glob.iglob(files) for files in f_filter] for f in files]

#---------------------------------------------------------------------------
#  Main prog begins here

def main(argv):

	foundfiles = False
	foundOurSelf = False

	Options = ProgramParameters()  #  Creates Options object for program varibles.

	InputFiles = Options.InputFiles

	files = []
	if len(InputFiles) == 0:
		# Get list of file OurSelf
		foundOurSelf = True
		if Options.RecursiveMode:
			files = RecursiveGlob(os.getcwd(), '/*.sym')
		else:
			files = glob.glob('*.sym')

	if len(files) > 0:
		foundfiles=True
		for file in files:
			InputFiles.append(file)

	if foundOurSelf and len(InputFiles) == 0:
		print 'No symbols found or specifed, try using --help'
	elif not foundfiles :
		if len(InputFiles) > 1:
			for file in InputFiles:
				if not os.path.isfile(file):
					print "Bad file name or option: " + file
					print Usage
	else:
		for symbol in InputFiles:
			ProcessSymbol(Options, symbol)

if __name__ == "__main__":
   main(sys.argv[1:])

# Done!

